### DISCUSSION TITLE
Where does the syscall(in gramine libos) implement stack pointer change?

### DISCUSSION THREAD

**Discussion by blossomin** [gramineproject/gramine#1527] [General]
Hello, 

I'm studying the source code of gramine libos syscall implementation. And I found that user program uses a "user" stack, while the libos and pal should share a "kernel" stack. When I checked the source code for how syscall is implemented and wanted to track the call stack for a syscall from user space to libos "kernel" space,  I found all the syscalls should go to this line("https://github.com/gramineproject/gramine/blob/master/libos/src/arch/x86_64/syscallas.S#L30). I scanned this assembly code, but I didn't find where the stack pointer is changed.

Does anyone know where this happenes? I searched in google, https://www.felixcloutier.com/x86/syscall.html says "The SYSCALL instruction does not save the stack pointer (RSP). If the OS system-call handler will change the stack pointer, it is the responsibility of software to save the previous value of the stack pointer. " 

Another point I don't understand: In general linux platform, the instruction SYSCALL should lift the privilege of the current program and start to execute code of kernel in RING 0. But in gramine libos context, the libos and user application execute in the same privilege level, RING 3.  How to correctly implement or change the semantics for SYSCALL?

Thanks in advance!

**Comment by dimakuv** [comment]
> I'm studying the source code of gramine libos syscall implementation. And I found that user program uses a "user" stack, while the libos and pal should share a "kernel" stack. When I checked the source code for how syscall is implemented and wanted to track the call stack for a syscall from user space to libos "kernel" space, I found all the syscalls should go to this line("https://github.com/gramineproject/gramine/blob/master/libos/src/arch/x86_64/syscallas.S#L30). I scanned this assembly code, but I didn't find where the stack pointer is changed.

But it is right there:
- Save "user" stack to register r11 (it will be saved in a special memory area a bit later, see next bullet point): https://github.com/gramineproject/gramine/blob/10c2668e35b11eacc0182c831f285113eeedc2b9/libos/src/arch/x86_64/syscallas.S#L40
- Register r11 (which now contains the "user" stack value) is saved in a special memory area: https://github.com/gramineproject/gramine/blob/10c2668e35b11eacc0182c831f285113eeedc2b9/libos/src/arch/x86_64/syscallas.S#L77
- Set the "kernel" stack: https://github.com/gramineproject/gramine/blob/10c2668e35b11eacc0182c831f285113eeedc2b9/libos/src/arch/x86_64/syscallas.S#L44

Finally, when restoring the "user" context, the "user" stack is also restored (from that special memory area where it was saved on syscall entry): https://github.com/gramineproject/gramine/blob/10c2668e35b11eacc0182c831f285113eeedc2b9/libos/src/arch/x86_64/syscallas.S#L266

---

> Another point I don't understand: In general linux platform, the instruction SYSCALL should lift the privilege of the current program and start to execute code of kernel in RING 0. But in gramine libos context, the libos and user application execute in the same privilege level, RING 3. How to correctly implement or change the semantics for SYSCALL?

Two points:
1. Gramine ships with a patched Glibc library (https://github.com/gramineproject/gramine/tree/master/subprojects/packagefiles/glibc-2.37). This Glibc library is patched for a single purpose: to replace SYSCALL instructions with function calls into Gramine, see https://github.com/gramineproject/gramine/blob/master/subprojects/packagefiles/glibc-2.37/gramine-syscall.patch. Obviously, applications running in Gramine must be instructed to use this patched Glibc library, which is achieved by the special `LD_LIBRARY_PATH` trick, see e.g. [this line](https://github.com/gramineproject/gramine/blob/10c2668e35b11eacc0182c831f285113eeedc2b9/CI-Examples/helloworld/helloworld.manifest.template#L7) together with [this line](https://github.com/gramineproject/gramine/blob/10c2668e35b11eacc0182c831f285113eeedc2b9/CI-Examples/helloworld/helloworld.manifest.template#L10).
2. If the application for some reason doesn't want/can use the patched Glibc library, then Gramine has two fallback mechanisms:
   a. On `gramine-direct` (no-SGX pal), we set up a seccomp filter in host Linux that catches SYSCALL instructions and redirects them back into Gramine. This approach is described well in the [academic paper on Gramine](https://www.chiachetsai.com/files/eurosys14.pdf).
   b. On `gramine-sgx` (SGX pal), we rely on the Intel SGX hardware itself -- it prohibits SYSCALL instructions and immediately redirects them into Gramine (by raising an Illegal Instruction exception). I could give you reference points in the code, if you'll want to get a closer look at this flow.

**Comment by blossomin** [comment]
Thanks, I understand it now.
> Save "user" stack to register r11 (it will be saved in a special memory area a bit later, see next bullet point):

 I thought that when it starts to execute [this code](https://github.com/gramineproject/gramine/blob/10c2668e35b11eacc0182c831f285113eeedc2b9/libos/src/arch/x86_64/syscallas.S#L30-L40), it already uses the "kernel" stack. (this is wrong).

> This Glibc library is patched for a single purpose: to replace SYSCALL instructions with function calls into Gramine, see https://github.com/gramineproject/gramine/blob/master/subprojects/packagefiles/glibc-2.37/gramine-syscall.patch. 
I see that all the "syscall" instruction is replaced by "GRAMINE_SYSCALL", which makes me wonder how "GRAMINE_SYSCALL" is implemented?

It seems that "GRAMINE_SYSCALL" is a [assembly macro](https://github.com/gramineproject/gramine/blob/10c2668e35b11eacc0182c831f285113eeedc2b9/libos/include/arch/x86_64/gramine_entry_api.h#L38-L44). In fact, I'm not familiar with assembly macro(I can understand basic assembly code), could you share more about this macro? Will these several assembly lines be inlined into glibc code?

Thanks again.

**Comment by blossomin** [comment]
I have another confusion point about EDMM inside the Gramine, although this is not related to this topic:

For SGX2 instruction EMODPE, it's optional to use EACCEPT, if I understand correctly. This claim can be supported by [these lines](https://github.com/gramineproject/gramine/blob/master/pal/src/host/linux-sgx/enclave_edmm.c#L54-L81). It first extends potential protection access rights, and if these pages have Read and Write protection access rights, then it doesn't need to restrict and so no EACCEPT is executed.

But for function to set page permission, [sgx_edmm_set_page_permissions](https://github.com/gramineproject/gramine/blob/master/pal/src/host/linux-sgx/enclave_edmm.c#L122-L145): if the expected permission is only extended one compared to current permission, we can only use EMODPE without EMODPR/EACCEPT, which might speedup the execution. So why not consider seperate this sgx_edmm_set_page_permissions into two, sgx_edmm_extend_page_permissions and sgx_edmm_restrict_page_permissions? Will it be hard to get the EPC access permission inside the enclave?

**Reply by dimakuv** [reply]
> But for function to set page permission, [sgx_edmm_set_page_permissions](https://github.com/gramineproject/gramine/blob/master/pal/src/host/linux-sgx/enclave_edmm.c#L122-L145): if the expected permission is only extended one compared to current permission, we can only use EMODPE without EMODPR/EACCEPT, which might speedup the execution. So why not consider seperate this sgx_edmm_set_page_permissions into two, sgx_edmm_extend_page_permissions and sgx_edmm_restrict_page_permissions? Will it be hard to get the EPC access permission inside the enclave?

Yes, this is a very good remark. Your understanding is correct.

Two main reasons:
1. The current code is simpler than what you suggest. The current code just blindly restricts the permissions first, and then extends them. This flow always succeeds.
2. The current code doesn't require to know the **currently set page permissions**. To write your proposed code, we would need to know the currently set page permissions, and this information is currently not easily available to the SGX EDMM code (this info is hidden in the LibOS layer).

Interestingly, we have a [PR that exposes this information to the SGX EDMM code](https://github.com/gramineproject/gramine/pull/1513), so maybe with that PR we'll indeed refactor the code as you suggest.

Obviously, the disadvantage of our current naive code is that this can be slow (we always perform `ocall_edmm_restrict_pages_perm()` which is an expensive operation, and for cases like `R` -> `RW` this ocall is not required).

**Reply by blossomin** [reply]
>The current code just blindly restricts the permissions first, and then extends them. This flow always succeeds.

In fact, I think the current code extends permission first and restricts it, [here](https://github.com/gramineproject/gramine/blob/10c2668e35b11eacc0182c831f285113eeedc2b9/pal/src/host/linux-sgx/enclave_edmm.c#L122-L126).

I checked another technical report from Intel([here](https://caslab.csl.yale.edu/workshops/hasp2016/HASP16-16.pdf)), in the section 3.3, they mentioned a protocol to extend page permission, and the second step needs to notify host os manager. However, for current gramine edmm support, I didn't see such a notification. Sure, I know that EMODPE doesn't need EACCEPT to synchronize TLB. I'm wondering whether the second step described here is necessary.
<img width="522" alt="截屏2023-09-03 10 34 45" src="https://github.com/gramineproject/gramine/assets/86922326/2f9ca7f3-9f1b-4217-b667-f130dfa777a3">

**Reply by dimakuv** [reply]
> > The current code just blindly restricts the permissions first, and then extends them. This flow always succeeds.
> 
> In fact, I think the current code extends permission first and restricts it, [here](https://github.com/gramineproject/gramine/blob/10c2668e35b11eacc0182c831f285113eeedc2b9/pal/src/host/linux-sgx/enclave_edmm.c#L122-L126).

Sorry, my bad. I made a typo. You are correct.

> I checked another technical report from Intel([here](https://caslab.csl.yale.edu/workshops/hasp2016/HASP16-16.pdf)), in the section 3.3, they mentioned a protocol to extend page permission, and the second step needs to notify host os manager. However, for current gramine edmm support, I didn't see such a notification. Sure, I know that EMODPE doesn't need EACCEPT to synchronize TLB. I'm wondering whether the second step described here is necessary.

Interesting. Yes, as far as I know, this second step is not required. But I will also ask other SGX experts, they should know better. Also maybe @kailun-qin @llly @rchatre know the answer?

**Reply by dimakuv** [reply]
> I checked another technical report from Intel([here](https://caslab.csl.yale.edu/workshops/hasp2016/HASP16-16.pdf)), in the section 3.3, they mentioned a protocol to extend page permission, and the second step needs to notify host os manager. However, for current gramine edmm support, I didn't see such a notification. Sure, I know that EMODPE doesn't need EACCEPT to synchronize TLB. I'm wondering whether the second step described here is necessary.

Ok, I remembered why there is no Step 2.

This is because in this function, we anyway [call `ocall_edmm_restrict_pages_perm()`](https://github.com/gramineproject/gramine/blob/aeb3fa952202a223453edd1d596b6ec0dc5d484f/pal/src/host/linux-sgx/enclave_edmm.c#L126C15-L126C45) which requests the system manager to reset page permissions in the page tables.

So, Step 2 becomes redundant, as the OS kernel will be instructed to reset page permissions anyway.

**Reply by blossomin** [reply]
> Sorry, my bad. I made a typo. You are correct.

In that report, they recommend to restrict first and then extend it at the end of section 3.3.
<img width="564" alt="截屏2023-09-03 11 50 52" src="https://github.com/gramineproject/gramine/assets/86922326/3cc4a2c5-a0f6-4a75-a2f4-3414489fcd9a">

Anyway, I guess there is no big difference.

**Reply by blossomin** [reply]
> This is because in this function, we anyway [call ocall_edmm_restrict_pages_perm()](https://github.com/gramineproject/gramine/blob/aeb3fa952202a223453edd1d596b6ec0dc5d484f/pal/src/host/linux-sgx/enclave_edmm.c#L126C15-L126C45) which requests the system manager to reset page permissions in the page tables.

Yes, in `sgx_edmm_set_page_permissions` function, there will be one _ocall_edmm_restrict_pages_perm()_. But [sgx_edmm_add_pages](https://github.com/gramineproject/gramine/blob/master/pal/src/host/linux-sgx/enclave_edmm.c#L41-L80) doesn't need to call _ocall_edmm_restrict_pages_perm_ sometimes. e.g. we want to add some pages with RWX permission.

**Reply by dimakuv** [reply]
> > This is because in this function, we anyway [call ocall_edmm_restrict_pages_perm()](https://github.com/gramineproject/gramine/blob/aeb3fa952202a223453edd1d596b6ec0dc5d484f/pal/src/host/linux-sgx/enclave_edmm.c#L126C15-L126C45) which requests the system manager to reset page permissions in the page tables.
> 
> Yes, in `sgx_edmm_set_page_permissions` function, there will be one _ocall_edmm_restrict_pages_perm()_. But [sgx_edmm_add_pages](https://github.com/gramineproject/gramine/blob/master/pal/src/host/linux-sgx/enclave_edmm.c#L41-L80) doesn't need to call _ocall_edmm_restrict_pages_perm_ sometimes. e.g. we want to add some pages with RWX permission.

Very good point. This indeed looks like a bug.

**Reply by dimakuv** [reply]
Actually, taking another look at the Linux SGX driver, I see this paragraph: https://lore.kernel.org/lkml/2617bf2b2d1e27ca1d0096e1192ae5896baf3f80.1652137848.git.reinette.chatre@intel.com/T/#m7c151a689f712eb88ae3409cef4d2eeb68a6ccab

> Add support for dynamically adding pages to an initialized enclave,
architecturally limited to RW permission at creation but allowed to
obtain RWX permissions after trusted enclave runs EMODPE. Add pages
via the page fault handler at the time an enclave address without a
backing enclave page is accessed, potentially directly reclaiming
pages if no free pages are available.

To me, this feels like there is **no need** to perform any additional OCALL IOCTL to the SGX driver. A simple `EMODPE` is sufficient.

@kailun-qin @llly @rchatre Could any of you confirm my understanding?

**Comment by dimakuv** [comment]
> It seems that "GRAMINE_SYSCALL" is a [assembly macro](https://github.com/gramineproject/gramine/blob/10c2668e35b11eacc0182c831f285113eeedc2b9/libos/include/arch/x86_64/gramine_entry_api.h#L38-L44). In fact, I'm not familiar with assembly macro(I can understand basic assembly code), could you share more about this macro? Will these several assembly lines be inlined into glibc code?

Yes, these several assembly lines will be inlined into Glibc code.

What this assembly snippet does it basically the following:
1. Put the [code right-after](https://github.com/gramineproject/gramine/blob/10c2668e35b11eacc0182c831f285113eeedc2b9/libos/include/arch/x86_64/gramine_entry_api.h#L42C32-L42C32) the ["jump to Gramine syscall entry" instruction](https://github.com/gramineproject/gramine/blob/10c2668e35b11eacc0182c831f285113eeedc2b9/libos/include/arch/x86_64/gramine_entry_api.h#L41) into [the RCX register](https://github.com/gramineproject/gramine/blob/10c2668e35b11eacc0182c831f285113eeedc2b9/libos/include/arch/x86_64/gramine_entry_api.h#L40). Recall that [RCX by convention contains the value to which the syscall will return after it is done](https://github.com/gramineproject/gramine/blob/10c2668e35b11eacc0182c831f285113eeedc2b9/libos/src/arch/x86_64/syscallas.S#L31).
2. Jump to Gramine syscall entry. This is done via the `jmp gs:24` instruction, which means "take the GS offset register value, [add 24](https://github.com/gramineproject/gramine/blob/10c2668e35b11eacc0182c831f285113eeedc2b9/libos/include/arch/x86_64/gramine_entry_api.h#L20) and jump to the resulting address". This address `gs + 24` is a special Gramine value: it is [`libos_tcb::libos_syscall_entry` value](https://github.com/gramineproject/gramine/blob/aeb3fa952202a223453edd1d596b6ec0dc5d484f/libos/include/libos_tcb.h#L21C26-L21C45), and `24` comes from [PAL TCB fields](https://github.com/gramineproject/gramine/blob/aeb3fa952202a223453edd1d596b6ec0dc5d484f/pal/include/arch/x86_64/pal_arch.h#L32) and [LibOS TCB fields](https://github.com/gramineproject/gramine/blob/aeb3fa952202a223453edd1d596b6ec0dc5d484f/libos/include/libos_tcb.h#L17). And this field points to our [asm function](https://github.com/gramineproject/gramine/blob/10c2668e35b11eacc0182c831f285113eeedc2b9/libos/src/arch/x86_64/syscallas.S#L30-L40).

**Reply by blossomin** [reply]
Thanks, I reviewed this part, PAL_TCB, and libos TCB. I have a better understanding now.

**Comment by blossomin** [comment]
Another question about Stack in libos:

I found there are two stack inside the libos thread, [first](https://github.com/gramineproject/gramine/blob/master/libos/include/libos_thread.h#L68) and [second](https://github.com/gramineproject/gramine/blob/master/libos/include/libos_thread.h#L127-L129). If I understand well, the first stack should be libos internal use, and the second is used by user application program?

**Reply by dimakuv** [reply]
Yes, your understanding is exactly correct.

**Reply by blossomin** [reply]
Thanks for you confirm.

I checked the thread implementation in gramine and found there are two _g_thread_list_, one in pal and the other in libos. Is the thread in Pal is one-to-one mapping to one thread in libos?

![截屏2023-09-05 09 17 30](https://github.com/gramineproject/gramine/assets/86922326/96ff1521-537f-44e5-b66d-16349c06503f)

As this figure shows, gramine allocates two stacks for its two threads, I think these two stacks are used by pal, but why do we need two seperate stack for pal and libos? 
And what is the usage of these two stacks, when is used?

**Reply by dimakuv** [reply]
> I checked the thread implementation in gramine and found there are two g_thread_list, one in pal and the other in libos. Is the thread in Pal is one-to-one mapping to one thread in libos?

Yes, it's one-to-one mapping.

> As this figure shows, gramine allocates two stacks for its two threads, I think these two stacks are used by pal, but why do we need two seperate stack for pal and libos?

Yes, the two stacks that are shown in the screenshot are used by the PAL. The enclave memory for these stacks is reserved at build time (this is a requirement of Intel SGX hardware).

The LibOS stacks are allocated on demand, and they are actually used at runtime (PAL stacks are used only at initialization time of each thread).

You are right that technically, there is no point in having two stacks for each thread in Gramine. We even have some comments about this:
- https://github.com/gramineproject/gramine/blob/94123cb244cf1b4dbfc0bfcb0805806ca6fcb1a5/libos/src/bookkeep/libos_thread.c#L210-L211
- https://github.com/gramineproject/gramine/issues/84 (first bullet)

> And what is the usage of these two stacks, when is used?

The PAL stack is used when the thread is created. It is the initial stack of the thread.

The LibOS stack is used at runtime. At some point in thread initialization, the thread switches from the initial PAL stack to the newly-allocated LibOS stack. After this point, the thread *always* uses the LibOS stack, and the PAL stack becomes useless.

**Reply by blossomin** [reply]
Hello again, 

Could you share some insight about the _sig_stack_? I checked the source code and found the _sig_stack_ should be used for exception handling, but the assembly code about exception handling in this file **enclave_ecall.S** is too complex to catch up for me. I just got some rough idea about it, could you share more details?

As for the stack usage, there are five stacks at least, 1. PAL stack, 2 PAL sig_stack, 3 libos internal stack and 4 application stack, 5 untrusted host stack

Let's just assume that the application program we want to run has only one thread(btw, do you know any method to restrict that the applictation only has one thread? mask the libos_syscall_clone? ), and the libos has already finished initialization and passed control to application program. So PAL stack won't be used anymore.
1) Normal ocall case: when the user application runs, it uses application stack. When it needs syscall from libos, it will change the stack to libos internal stack.
And if this syscall or functions inside libos needs ocall, it will change the stack to untrusted host stack in **enclave_ocall.S** file. When this ocall finishes, restore the stack to libos internal stack. when this syscall completes, restore the stack to application stack.

2) AEX case: I don't know how AEX happens, could you share anything about AEX? How does the AEX/exception use _sig_stack_?

Thanks!

**Reply by dimakuv** [reply]
Sorry, discussing the exception handling will take too much time.

I suggest to google for "2 phase exception handling in SGX". You can also read the paper https://n.ethz.ch/~sshivaji/publications/smashex_ccs21.pdf, though it doesn't give the whole picture of the exception handling mechanism.

Some code points which can give you a start:
- https://github.com/gramineproject/gramine/blob/8f65e6ad7444410d576a63fd0e5f443ced82830c/pal/src/host/linux-sgx/host_exception.c#L165
- https://github.com/gramineproject/gramine/blob/8f65e6ad7444410d576a63fd0e5f443ced82830c/pal/src/host/linux-sgx/host_entry.S#L125
- https://github.com/gramineproject/gramine/blob/8f65e6ad7444410d576a63fd0e5f443ced82830c/pal/src/host/linux-sgx/enclave_entry.S#L88
- https://github.com/gramineproject/gramine/blob/8f65e6ad7444410d576a63fd0e5f443ced82830c/pal/src/host/linux-sgx/enclave_entry.S#L191

> ... btw, do you know any method to restrict that the applictation only has one thread? mask the libos_syscall_clone?

Well, if you're willing to modify Gramine to do it, then yes, just return `-EPERM` at the very beginning of `libos_syscall_clone()`. This will prevent the application to create any new threads.

> Normal ocall case: when the user application runs, it uses application stack. When it needs syscall from libos, it will change the stack to libos internal stack. And if this syscall or functions inside libos needs ocall, it will change the stack to untrusted host stack in enclave_ocall.S file. When this ocall finishes, restore the stack to libos internal stack. when this syscall completes, restore the stack to application stack.

Yes, this description is correct.

**Reply by blossomin** [reply]
Hello, I have a basic understanding for exception handling inside the gramine/sgx2, but I found several confused points during my inspeciton. 
1. a naive question: why **cld** instruction needs to go to **async_exit_pointer**? Because this instruction needs to clear DF in EFLAGS,  it needs to execute outside enclave?

![截屏2023-09-13 11 57 06](https://github.com/gramineproject/gramine/assets/86922326/8c7ee259-236d-4913-8358-b04cd83e0153)
2. %RCX register is set to the address of **async_exit_pointer** [here](https://github.com/gramineproject/gramine/blob/master/pal/src/host/linux-sgx/host_entry.S#L40), I don't understand why it doesn't execute line 71(lock incq %gs:PAL_HOST_TCB_AEX_CNT), which should the first instruction of **async_exit_pointer**, and directly jump to line 75 after executing **cld**.
![截屏2023-09-13 12 00 11](https://github.com/gramineproject/gramine/assets/86922326/c4eda7e6-7325-46fe-8e91-6233276f85a0)
3. why is %RAX 3 here? I don't know the detail but I guess %RAX is set to 3 when the control flow jumps to **async_exit_pointer**, could you share any hint?

Thanks

**Reply by dimakuv** [reply]
> a naive question: why cld instruction needs to go to async_exit_pointer? Because this instruction needs to clear DF in EFLAGS, it needs to execute outside enclave?

That's not correct. The `cld` instruction doesn't exit the enclave and doesn't jump to `async_exit_pointer`.

In reality, what you observe is a "glitch" in GDB execution -- GDB periodically "looses track" of the SGX enclave execution, and GDB kinda "restarts" from the `async_exit_pointer` place in code. This is an unfortunate bug in GDB/SGX, and we don't know how to fix it.

So just ignore these "jumps to AEX untrusted landing pad" when you debug the enclave code.

> 2. %RCX register is set to the address of async_exit_pointer [here](https://github.com/gramineproject/gramine/blob/master/pal/src/host/linux-sgx/host_entry.S#L40), I don't understand why it doesn't execute line 71(lock incq %gs:PAL_HOST_TCB_AEX_CNT), which should the first instruction of async_exit_pointer, and directly jump to line 75 after executing cld.

Same explanation -- this is a bug in GDB execution. You're right that actually line 71 should have been executed, but GDB for whatever reason jumped over it and is now located on line 75. Again, just ignore these glitches.

> 3. why is %RAX 3 here? I don't know the detail but I guess %RAX is set to 3 when the control flow jumps to async_exit_pointer, could you share any hint?

Now that's a good question. You must look into the Intel SDM documentation, in particular, Table 37-1 in Volume 3D, Chapter 37.3.1 "Processor Synthetic State on Asynchronous Enclave Exit".

The documentation is very clear:
- RAX -- value is 3 (ENCLU[3] is ERESUME).

**Comment by blossomin** [comment]
Morning,

I have some confused points for edmm inside gramine. If we enable edmm in the gramine manifest, the enclave image signing and initialization will skip the **free** section, which should be the heap memory of pal loader. It hurts the overall performance although it can accelerate enclave page adding. 

To my understanding, we can preallocate the **free** section like disabling edmm, for example, we can allocate same size free section like edmm disabled, of course, we can set these page as ---(not readable, not writable, not executable) and just use edmm property to change page permission as request when we really want this page. So I assume this can somehow optimize.

However, as I tested on my side, the libos can successfully initialize, and start executing user program, but it will wait forever on futex syscall. I compared the execution log for 1)edmm disabled version and 2)edmm enable preallocate. I found they issued different syscalls. disable edmm version won't call futex syscall,. before this different point, all the logs looks similar, except some hash values or memory addresses.

Do you have any idea about this?

**Reply by dimakuv** [reply]
> To my understanding, we can preallocate the free section like disabling edmm, for example, we can allocate same size free section like edmm disabled, of course, we can set these page as ---(not readable, not writable, not executable) and just use edmm property to change page permission as request when we really want this page. So I assume this can somehow optimize.

Yes, this looks like a reasonable thing to explore. I'm not sure if this will produce good performance optimizations, but you can always try.

> However, as I tested on my side, the libos can successfully initialize, and start executing user program, but it will wait forever on futex syscall. I compared the execution log for 1)edmm disabled version and 2)edmm enable preallocate. I found they issued different syscalls. disable edmm version won't call futex syscall,. before this different point, all the logs looks similar, except some hash values or memory addresses.

This looks very much like a problem with a lock. Recall that locks (mutexes) are implemented like this:
1. First try to access an atomic variable (stored in a memory region shared by all threads). If the variable is `0`, then grab the lock and set this variable to `1` ("lock acquired").
2. Second, if the atomic variable is `1` ("lock already acquired"), then wait until this variable is set to `0`. The "wait" is performed using the `futex()` syscall.

So apparently you have a bug at step 1. In your "EDMM enable preallocate" case that you programmed, apparently some memory region is NOT correctly initialized, and the atomic variable most probably contains some garbage value (like `123`). So the lock goes to step 2 and waits on a `futex()`, even though in reality the lock is supposed to be in the "lock available" state.

This is just my theory, but I have a strong suspicion that's what happens. In other words, you forgot to zero-out the memory that you EDMM-preallocated.

**Reply by blossomin** [reply]
Yes, it works after I zero all the pages before real allocation, athough I just create the enclave image with RWX free section. It will be better to set the free section as ---, in case any security issue.

**Reply by dimakuv** [reply]
> Yes, it works after I zero all the pages before real allocation, athough I just create the enclave image with RWX free section. It will be better to set the free section as ---, in case any security issue.

Well, you can do it. You just need to change the page protections during `PalVirtualMemoryAlloc()`, by calling appropriate EDMM flows. I thought that was the whole point of your modification to Gramine?

**Reply by blossomin** [reply]
In fact, I added another option in manifest, edmm_preallocate. Here is my modification: https://github.com/blossomin/gramine_sandbox/commit/3e78d53100d1102357725d2eeef8cebf200befd7

And I want to change this line: [free_area->prot](https://github.com/blossomin/gramine_sandbox/blob/preallocate/pal/src/host/linux-sgx/host_main.c#L480) to: free_area->prot = 0;, of course, I need to change the corresponding sgx_sign.py file. but after this modification, it seems stuck in enclave initialization.

**Reply by dimakuv** [reply]
Sorry, @blossomin, this is getting too low-level. I won't be able to spend more time on helping you with the code :(

**Reply by blossomin** [reply]
Yeah, thanks for your discussion, which helps me a lot. Of course, I will try myself when I'm free. At least I have a runnable version now. And I have a default design: allocate all the free section pages with RWX, and after the enclave initialization, change the pages to --- at the beginning of enclave start. Although there might be other issues, this could be a solution, not that elegant.

Thanks again!

Also, maybe this can be a possible optimization for libos with edmm due to the slow SGX2 instructions.
